### PR TITLE
Work around LLVM issues with explicit register in inline asm

### LINKED ISSUE #74658: Segfault when using an ymm register as an out parameter in an asm! block.
### Code

The following code segfaults when compiled (`rustc asm-avx.rs -C target-cpu=native`)

```Rust
#![feature(asm)]

use std::alloc;

#[cfg(target_feature = "avx")]
fn main() {
    unsafe {
        let layout = alloc::Layout::from_size_align_unchecked(32, 32);
        let mem = alloc::alloc_zeroed(layout);
        asm!(
             "vmovdqa ymm0, [{source}]",
             source = in(reg) mem,
             out("ymm0") _,
        );
    }
}
```

Omitting (incorrectly) the register for the inputs and outputs does not cause a segfault.

```rust
#![feature(asm)]

use std::alloc;

#[cfg(target_feature = "avx")]
fn main() {
    unsafe {
        let layout = alloc::Layout::from_size_align_unchecked(32, 32);
        let mem = alloc::alloc_zeroed(layout);
        asm!(
             "vmovdqa ymm0, [{source}]",
             source = in(reg) mem,
        );
    }
}
```

`xmm` registers work fine even when they appear in outputs.

```rust
#![feature(asm)]

use std::alloc;

fn main() {
    unsafe {
        let layout = alloc::Layout::from_size_align_unchecked(16, 16);
        let mem = alloc::alloc_zeroed(layout);
        asm!(
             "movdqa xmm0, [{source}]",
             source = in(reg) mem,
             out("xmm0") _,
        );
    }
}
```

### Meta

`rustc --version --verbose`:
```
rustc 1.47.0-nightly (8ad7bc3f4 2020-07-21)
binary: rustc
commit-hash: 8ad7bc3f428300aee6764f6e23527e19eb235e81
commit-date: 2020-07-21
host: x86_64-apple-darwin
release: 1.47.0-nightly
LLVM version: 10.0
```

### Error output

There is no output, `rustc` segfaults before emitting any.

### PR DESCRIPTION
**Pull request by Amanieu**
Fixes #74658

### CODE REVIEW

File: src/librustc_codegen_llvm/asm.rs [hunk 1]
```diff
@@ -410,10 +413,80 @@ fn inline_asm_call(
     }
 }
 
+/// If the register is an xmm/ymm/zmm register then return its index.
+fn xmm_reg_index(reg: InlineAsmReg) -> Option<u32> {
+    match reg {
+        InlineAsmReg::X86(reg)
+            if reg as u32 >= X86InlineAsmReg::xmm0 as u32
+                && reg as u32 <= X86InlineAsmReg::xmm15 as u32 =>
```
Comment (lzutao) [inline src/librustc_codegen_llvm/asm.rs:421]:
```suggestion
        InlineAsmReg::X86(reg)
            if RangeInclusive::new(X86InlineAsmReg::xmm0, X86InlineAsmReg::xmm15).contains(&reg) =>
```

Comment (Amanieu) [reply src/librustc_codegen_llvm/asm.rs:421]: Does `RangeInclusive` work on enums?

Comment (lzutao) [reply src/librustc_codegen_llvm/asm.rs:421]: If it implement PartialOrd and/or PartialEq I think.

Comment (eddyb) [reply src/librustc_codegen_llvm/asm.rs:421]: You can just write `(X86InlineAsmReg::xmm0..=X86InlineAsmReg::xmm15).contains(&reg)`, too.

**Comment by eddyb** [comment]
Looks fine I think, but I'd suggest avoiding waiting on reviews from me where possible.

r? @nagisa

**Comment by nagisa** [comment]
This mentions LLVM issues and that this is a workaround. A couple of questions: is this really a LLVM issue (besides the fact that LLVM fires an assertion rather than diagnostic callback)? If so, is there a report upstream? If not, could this PR be retitled to something else (just so that it is easier to search for/understand in the future).

Otherwise @bors r+

**Comment by Amanieu** [comment]
I'm not sure it is a bug in LLVM but rather just LLVM being very picky about what types you can use with a particular register type. This PR adjusts the types in rustc to match what LLVM expects.

**Pull request merged**